File: Integraupt-Backend/BackendHorarios/src/main/java/com/horarios/service/HorarioService.java
package com.horarios.service;

import com.horarios.dto.ActualizarOcupacionRequest;
import com.horarios.dto.HorarioDiaResponse;
import com.horarios.dto.HorarioRequest;
import com.horarios.dto.HorarioResponse;
import com.horarios.dto.HorarioSemanalResponse;
import com.horarios.mapper.HorarioMapper;
import com.horarios.model.DiaSemana;
import com.horarios.model.Espacio;
import com.horarios.model.Horario;
import com.horarios.repository.BloqueHorarioRepository;
import com.horarios.repository.EspacioRepository;
import com.horarios.repository.HorarioRepository;
import jakarta.transaction.Transactional;
import org.springframework.stereotype.Service;
import org.springframework.data.domain.Sort;
import com.horarios.model.BloqueHorario;

import java.time.format.DateTimeFormatter;
import java.util.Arrays;
import java.util.EnumMap;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@Service
@Transactional
public class HorarioService {

    private static final DateTimeFormatter HORA_FORMATTER = DateTimeFormatter.ofPattern("HH:mm");
    private final HorarioRepository horarioRepository;
    private final EspacioRepository espacioRepository;
    private final BloqueHorarioRepository bloqueHorarioRepository;

    public HorarioService(HorarioRepository horarioRepository,
                          EspacioRepository espacioRepository,
                          BloqueHorarioRepository bloqueHorarioRepository) {
        this.horarioRepository = horarioRepository;
        this.espacioRepository = espacioRepository;
        this.bloqueHorarioRepository = bloqueHorarioRepository;
    }

    public List<HorarioResponse> listarTodos() {
        return horarioRepository.findAll().stream()
                .map(HorarioMapper::toResponse)
                .toList();
    }

    public Optional<HorarioResponse> buscarPorId(Integer id) {
        return horarioRepository.findById(id).map(HorarioMapper::toResponse);
    }

    public List<HorarioResponse> listarPorEspacio(Integer espacioId) {
        return horarioRepository.findByEspacioIdOrderByBloqueOrdenAscDiaSemanaAsc(espacioId).stream()
                .map(HorarioMapper::toResponse)
                .toList();
    }

    public List<HorarioResponse> listarPorDia(String diaSemana) {
        DiaSemana dia = DiaSemana.fromNombre(diaSemana);
        return horarioRepository.findByDiaSemanaOrderByBloqueOrdenAsc(dia).stream()
                .map(HorarioMapper::toResponse)
                .toList();
    }

    public List<HorarioResponse> listarPorOcupacion(boolean ocupado) {
        return horarioRepository.findByOcupado(ocupado).stream()
                .map(HorarioMapper::toResponse)
                .toList();
    }

    public List<HorarioSemanalResponse> obtenerHorarioSemanalPorEspacio(Integer espacioId) {
        List<Horario> horarios = horarioRepository.findByEspacioIdOrderByBloqueOrdenAscDiaSemanaAsc(espacioId);

        Map<Integer, Map<DiaSemana, Boolean>> estadosPorBloque = new HashMap<>();
        horarios.forEach(horario -> {
            BloqueHorario bloque = horario.getBloque();
            DiaSemana dia = horario.getDiaSemana();

            if (bloque == null || bloque.getId() == null || dia == null) {
                return;
            }
            estadosPorBloque
                    .computeIfAbsent(bloque.getId(), id -> new EnumMap<>(DiaSemana.class))
                    .put(dia, horario.isOcupado());
        });

        return bloqueHorarioRepository.findAll(Sort.by(Sort.Direction.ASC, "orden")).stream()
                .map(bloque -> {
                    Map<DiaSemana, Boolean> estados = estadosPorBloque.getOrDefault(bloque.getId(), Map.of());

                    List<HorarioDiaResponse> dias = Arrays.stream(DiaSemana.values())
                            .map(dia -> new HorarioDiaResponse(dia.getNombre(), estados.getOrDefault(dia, false)))
                            .toList();

                    return new HorarioSemanalResponse(
                            bloque.getId(),
                            Optional.ofNullable(bloque.getNombre()).filter(nombre -> !nombre.isBlank())
                                    .orElseGet(() -> String.format("Bloque %d", bloque.getId())),
                            Optional.ofNullable(bloque.getHoraInicio())
                                    .map(HORA_FORMATTER::format)
                                    .orElse("--:--"),
                            Optional.ofNullable(bloque.getHoraFin())
                                    .map(HORA_FORMATTER::format)
                                    .orElse("--:--"),
                            dias
                    );

                })
                .toList();
    }

    public HorarioResponse crearHorario(HorarioRequest request) {
        Horario horario = new Horario();
        horario.setEspacio(obtenerEspacio(request.espacioId()));
        horario.setBloque(bloqueHorarioRepository.findById(request.bloqueId())
                .orElseThrow(() -> new IllegalArgumentException("Bloque horario no encontrado")));
        horario.setDiaSemana(DiaSemana.fromNombre(request.diaSemana()));
        horario.setOcupado(Boolean.TRUE.equals(request.ocupado()));
        Horario guardado = horarioRepository.save(horario);
        return HorarioMapper.toResponse(guardado);
    }

    public HorarioResponse actualizarHorario(Integer id, HorarioRequest request) {
        Horario horario = horarioRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Horario no encontrado"));
        horario.setEspacio(obtenerEspacio(request.espacioId()));
        horario.setBloque(bloqueHorarioRepository.findById(request.bloqueId())
                .orElseThrow(() -> new IllegalArgumentException("Bloque horario no encontrado")));
        horario.setDiaSemana(DiaSemana.fromNombre(request.diaSemana()));
        if (request.ocupado() != null) {
            horario.setOcupado(request.ocupado());
        }
        return HorarioMapper.toResponse(horario);
    }

    public void eliminarHorario(Integer id) {
        if (!horarioRepository.existsById(id)) {
            throw new IllegalArgumentException("Horario no encontrado");
        }
        horarioRepository.deleteById(id);
    }

    public HorarioResponse actualizarOcupacion(Integer id, ActualizarOcupacionRequest request) {
        Horario horario = horarioRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Horario no encontrado"));
        horario.setOcupado(request.ocupado());
        return HorarioMapper.toResponse(horario);
    }

    private Espacio obtenerEspacio(Integer espacioId) {
        return espacioRepository.findById(espacioId)
                .orElseThrow(() -> new IllegalArgumentException("Espacio no encontrado"));
    }
}

File: integraupt-web/Integraupt-Backend/backend-horarios-laravel/app/Services/HorarioService.php
<?php

namespace App\Services;

use App\Enums\DiaSemana;
use App\Models\BloqueHorario;
use App\Models\Espacio;
use App\Models\Horario;
use Carbon\Carbon;

/**
 * Servicio de negocio para Horarios.
 * Replica: com.horarios.service.HorarioService
 */
class HorarioService
{
    /**
     * Listar todos los horarios.
     * Replica: listarTodos()
     */
    public function listarTodos(): array
    {
        $horarios = Horario::with(['espacioRelation', 'bloqueRelation'])->get();
        return $horarios->map(fn($h) => $this->mapToResponse($h))->toArray();
    }

    /**
     * Buscar horario por ID.
     * Replica: buscarPorId(Integer id)
     */
    public function buscarPorId(int $id): ?array
    {
        $horario = Horario::with(['espacioRelation', 'bloqueRelation'])->find($id);
        return $horario ? $this->mapToResponse($horario) : null;
    }

    /**
     * Listar horarios por espacio, ordenados por bloque.orden ASC, diaSemana ASC.
     * Replica: listarPorEspacio(Integer espacioId)
     */
    public function listarPorEspacio(int $espacioId): array
    {
        $horarios = Horario::with(['espacioRelation', 'bloqueRelation'])
            ->where('horarios.espacio', $espacioId)
            ->join('bloqueshorarios', 'horarios.bloque', '=', 'bloqueshorarios.IdBloque')
            ->orderBy('bloqueshorarios.Orden', 'asc')
            ->orderBy('horarios.diaSemana', 'asc')
            ->select('horarios.*')
            ->get();

        return $horarios->map(fn($h) => $this->mapToResponse($h))->toArray();
    }

    /**
     * Listar horarios por día de la semana, ordenados por bloque.orden ASC.
     * Replica: listarPorDia(String diaSemana)
     */
    public function listarPorDia(string $diaSemana): array
    {
        $dia = DiaSemana::fromNombre($diaSemana);

        $horarios = Horario::with(['espacioRelation', 'bloqueRelation'])
            ->where('horarios.diaSemana', $dia->value)
            ->join('bloqueshorarios', 'horarios.bloque', '=', 'bloqueshorarios.IdBloque')
            ->orderBy('bloqueshorarios.Orden', 'asc')
            ->select('horarios.*')
            ->get();

        return $horarios->map(fn($h) => $this->mapToResponse($h))->toArray();
    }

    /**
     * Listar horarios por estado de ocupación.
     * Replica: listarPorOcupacion(boolean ocupado)
     */
    public function listarPorOcupacion(bool $ocupado): array
    {
        $horarios = Horario::with(['espacioRelation', 'bloqueRelation'])
            ->where('ocupado', $ocupado)
            ->get();

        return $horarios->map(fn($h) => $this->mapToResponse($h))->toArray();
    }

    /**
     * Obtener horario semanal completo por espacio.
     * Genera una matriz bloques × días con estado de ocupación.
     * Replica: obtenerHorarioSemanalPorEspacio(Integer espacioId)
     */
    public function obtenerHorarioSemanalPorEspacio(int $espacioId): array
    {
        // Obtener horarios del espacio
        $horarios = Horario::where('espacio', $espacioId)->get();

        // Indexar estados por bloqueId → diaSemana → ocupado
        $estadosPorBloque = [];
        foreach ($horarios as $horario) {
            $bloqueId = $horario->bloque;
            $dia = $horario->diaSemana;

            if ($bloqueId === null || $dia === null) {
                continue;
            }

            $estadosPorBloque[$bloqueId][$dia] = (bool) $horario->ocupado;
        }

        // Obtener todos los bloques ordenados
        $bloques = BloqueHorario::orderBy('Orden', 'asc')->get();
        $diasOrdenados = DiaSemana::ordered();

        return $bloques->map(function ($bloque) use ($estadosPorBloque, $diasOrdenados) {
            $estados = $estadosPorBloque[$bloque->IdBloque] ?? [];

            $dias = array_map(function ($dia) use ($estados) {
                return [
                    'diaSemana' => $dia->value,
                    'ocupado' => $estados[$dia->value] ?? false,
                ];
            }, $diasOrdenados);

            $nombre = !empty(trim($bloque->Nombre ?? ''))
                ? $bloque->Nombre
                : sprintf('Bloque %d', $bloque->IdBloque);

            $horaInicio = $bloque->HoraInicio
                ? Carbon::parse($bloque->HoraInicio)->format('H:i')
                : '--:--';

            $horaFin = $bloque->HoraFinal
                ? Carbon::parse($bloque->HoraFinal)->format('H:i')
                : '--:--';

            return [
                'bloqueId' => $bloque->IdBloque,
                'bloqueNombre' => $nombre,
                'horaInicio' => $horaInicio,
                'horaFin' => $horaFin,
                'dias' => $dias,
            ];
        })->toArray();
    }

    /**
     * Crear un nuevo horario.
     * Replica: crearHorario(HorarioRequest request)
     */
    public function crearHorario(array $data): array
    {
        $espacio = $this->obtenerEspacio($data['espacioId']);
        $bloque = BloqueHorario::findOrFail($data['bloqueId']);
        $dia = DiaSemana::fromNombre($data['diaSemana']);

        $horario = new Horario();
        $horario->espacio = $espacio->IdEspacio;
        $horario->bloque = $bloque->IdBloque;
        $horario->diaSemana = $dia->value;
        $horario->ocupado = !empty($data['ocupado']);
        $horario->save();

        $horario->load(['espacioRelation', 'bloqueRelation']);
        return $this->mapToResponse($horario);
    }

    /**
     * Actualizar un horario existente.
     * Replica: actualizarHorario(Integer id, HorarioRequest request)
     */
    public function actualizarHorario(int $id, array $data): array
    {
        $horario = Horario::findOrFail($id);

        $espacio = $this->obtenerEspacio($data['espacioId']);
        $bloque = BloqueHorario::findOrFail($data['bloqueId']);
        $dia = DiaSemana::fromNombre($data['diaSemana']);

        $horario->espacio = $espacio->IdEspacio;
        $horario->bloque = $bloque->IdBloque;
        $horario->diaSemana = $dia->value;

        if (isset($data['ocupado'])) {
            $horario->ocupado = $data['ocupado'];
        }

        $horario->save();
        $horario->load(['espacioRelation', 'bloqueRelation']);
        return $this->mapToResponse($horario);
    }

    /**
     * Eliminar un horario.
     * Replica: eliminarHorario(Integer id)
     */
    public function eliminarHorario(int $id): void
    {
        $horario = Horario::find($id);
        if (!$horario) {
            throw new \InvalidArgumentException('Horario no encontrado');
        }
        $horario->delete();
    }

    /**
     * Actualizar solo el estado de ocupación.
     * Replica: actualizarOcupacion(Integer id, ActualizarOcupacionRequest request)
     */
    public function actualizarOcupacion(int $id, bool $ocupado): array
    {
        $horario = Horario::findOrFail($id);
        $horario->ocupado = $ocupado;
        $horario->save();

        $horario->load(['espacioRelation', 'bloqueRelation']);
        return $this->mapToResponse($horario);
    }

    /**
     * Mapear un Horario Eloquent a la estructura de respuesta JSON.
     * Replica: HorarioMapper.toResponse(Horario horario)
     */
    private function mapToResponse(Horario $horario): array
    {
        $espacio = $horario->espacioRelation;
        $bloque = $horario->bloqueRelation;

        $horaInicio = $bloque && $bloque->HoraInicio
            ? Carbon::parse($bloque->HoraInicio)->format('H:i')
            : null;

        $horaFin = $bloque && $bloque->HoraFinal
            ? Carbon::parse($bloque->HoraFinal)->format('H:i')
            : null;

        return [
            'id' => $horario->IdHorario,
            'espacioId' => $espacio?->IdEspacio,
            'espacioNombre' => $espacio?->Nombre,
            'espacioCodigo' => $espacio?->Codigo,
            'bloqueId' => $bloque?->IdBloque,
            'bloqueNombre' => $bloque?->Nombre,
            'horaInicio' => $horaInicio,
            'horaFin' => $horaFin,
            'diaSemana' => $horario->diaSemana,
            'ocupado' => (bool) $horario->ocupado,
        ];
    }

    /**
     * Obtener espacio o lanzar excepción.
     * Replica: obtenerEspacio(Integer espacioId)
     */
    private function obtenerEspacio(int $espacioId): Espacio
    {
        $espacio = Espacio::find($espacioId);
        if (!$espacio) {
            throw new \InvalidArgumentException('Espacio no encontrado');
        }
        return $espacio;
    }
}
